### DISCUSSION TITLE
Implementing "run suite" in adapters

### DISCUSSION THREAD

**Discussion by rouge8** [nvim-neotest/neotest#221] [Q&A] [answered]
I wrote [neotest-rust](https://github.com/rouge8/neotest-rust) and I'm wondering what I need to implement for `require("neotest").run.run({ suite = true })` to work correctly. I haven't done anything special to handle this, but it looks like it already runs all the tests but only updates the test summary for the current file. When I open other files, the test summary updates for them too though.

For example, when I open https://github.com/tokio-rs/tokio/blob/master/tokio/tests/sync_rwlock.rs and run `:lua require("neotest").run.run({ suite = true })`, I see this in the test summary:

![SCR-20230317-osj](https://user-images.githubusercontent.com/237005/226074371-cbd1023e-f020-4fc8-871b-9b068e6e3dac.png)

When I open another file, it updates the summary (though the file itself is marked with a `?` rather than a check):

![SCR-20230317-otc](https://user-images.githubusercontent.com/237005/226074415-ed22f073-a364-45cd-bdfb-3f1528ce131c.png)

**Comment by rcarriga** [comment, marked as answer]
My guess here is that you have discovery disabled. The file you see is discovered because it's in an open buffer. I'm assuming that neotest-rust doesn't filter results for known files when running the suite so neotest has the results for all files but doesn't have those files in the tree so they're not shown. Enabling discovery will show all results

**Reply by rcarriga** [reply]
The reason for the `?` mark on the file is because neotest-rust isn't returning a result for the file ID itself, which neotest will infer if it's known at the time results are returned

**Reply by rouge8** [reply]
Ah right, I forgot about discovery! I disabled that a long time ago for performance reasons… 

Thanks for the pointer about file IDs, I’ll look at some of the other adapters to understand what to do there.

**Reply by rcarriga** [reply]
If the underlying test runner doesn't return results for files you could use similar login to neotest itself but it's a bit complex. It's implemented in this function
https://github.com/nvim-neotest/neotest/blob/bbbfa55d850f1aaa6707ea85fb5230ac866459c6/lua/neotest/client/runner.lua#L217
